[USM] add decoding of encrypted HTTP2 traffic

Yumasi opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

This PR makes USM able to decode the HTTP2 traffic from our TLS implementation.


### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

The PR includes unit tests for it.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

guyarb committed:
usm: http2: Moved common http2 methods to a new file

Moved 'is_interesting_static_entry', 'is_static_table_entry', 'http2_fetch_stream', 'get_dynamic_counter',
'parse_field_indexed', 'update_path_size_telemetry', 'handle_end_of_stream', 'format_http2_frame_header',
and 'reset_frame' to a new file 'decoding-common.h' that can be shared between http2-tls and http2-socket-filter
implementations. Those functions are agnostic to ctx object and are not tied to a specific program

guyarb committed:
usm: http2: Added new fields for tls_dispatcher_arguments_t

Added 'len' and 'off' fields which supposed to be equivalent for skb_info_t, and allow us to scan
the payload over multiple tail-calls like we do in the socket-filter programs.
The current commit has no actual impact, as it just adds the fields without using them

guyarb committed:
usm: http2: Added tls_http2_iterations map

The new map is the TLS version of http2_iterations and used to preserve state
between TLS tail-calls for http2 programs

guyarb committed:
usm: ebpf: utils: Introduce READ_USER_BUFFER macro with option to disable telemetry

The HTTP2 programs are very complicated for the verifer, and the telemetry versions of bpf_probe_read_user
adds irrelevant complexity. Thus the new macro should allow us to decide if we want telemetry or not

guyarb committed:
usm: http2: Added tls versions for http2 methods

This commit duplicates the methods in 'decoding.h' and modifies the functions to their tls version.
The new functions have 'tls_' prefix, and does use bpf_probe_read_user instead of
bpf_skb_load_bytes. Also the new methods use tls_dispatcher_arguments_t struct, rather
than using 'struct __sk_buff' and 'skb_info_t'

guyarb committed:
usm: http2: Added entrypoints for HTTP2-TLS

For every socket-filter program we have in http2 monitoring, we add a matching uprobe entrypoint.
Currently the entrypoints are empty and acts as placeholders.
Note - we need to add an additional entrypoint to handle the termination of the connection
as unlike socket-filter, the termination is another hook

guyarb committed:
usm: http2: Implement http2 termination program

Implemented the equivalent of cleanup logic (originally in socket__http2_handle_first_frame) in
uprobe entry.

guyarb committed:
usm: http2: Implement http2 tls handle first frame

Equivalent to socket__http2_handle_first_frame without connection termination logic which was extracted

guyarb committed:
usm: http2: Implement http2 tls filter entrypoint

Equivalent to socket__http2_filter

guyarb committed:
usm: http2: Implement uprobe__http2_tls_headers_parser

Implement the matching version of socket__http2_headers_parser.
Note - We had to declare HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL,
HTTP2_TLS_MAX_TAIL_CALLS_FOR_HEADERS_PARSER and HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER, to handle
complexity problems with the verifier

guyarb committed:
usm: http2: Implement uprobe__http2_tls_eos_parser

Implement the matching version of socket__http2_eos_parser.

Yumasi committed:
usm: http2: tests: add grpc-tls tests

guyarb committed:
usm: http2: tls: Support kIndexPath and kEmptyPath for literal values

We handled the cases were both key 4 and key 5 were used in literal key for path

guyarb committed:
usm: http2: tls: Add support for literal header never indexed and literal header without indexing

Making a support for parse and process literal header never indexed (prefix 0001) and
literal header without indexing (prefix 0000) in the headers processing

guyarb committed:
usm: http2: tls: Add support for getting frame-header and frame-payload
in different calls

We've seen libraries that read (write) the frame header separately from the frame payload
thus we need to cache the frame header and wait for the next call

guyarb committed:
usm: http2: tls: Added http2-go-tls tests

Extended the already existing http-go-tls tests to run also for http2

guyarb committed:
usm: http2: Move grpc/monitor_test to tracer package

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-defs.h:
```diff
@@ -26,6 +26,13 @@
 #define HTTP2_MAX_FRAMES_TO_FILTER  120
 
+// Represents the maximum number of frames we'll process in a single tail call in `uprobe__http2_tls_headers_parser` program.
+#define HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL 10
+// Represents the maximum number of tail calls to process headers frames.
+// Currently we have up to 120 frames in a packet, thus 12 (12*10 = 120) tail calls is enough.
+#define HTTP2_TLS_MAX_TAIL_CALLS_FOR_HEADERS_PARSER 12
```
Can we make the same values for both plain and TLS? in my eyes it can be really confusing in the future to handle issues

guyarb replied:
No we can't, as we have stricter limitation in the uprobe version, that should not affect the socket-filter ones

amitslavin replied:
Yes I see, that does make sense

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -667,6 +667,97 @@ int uprobe__http2_tls_headers_parser(struct pt_regs *ctx) {
+    const __u32 zero = 0;
+
+    tls_dispatcher_arguments_t dispatcher_args_copy;
+    // We're not calling fetch_dispatching_arguments as, we need to modify the `off` field of skb_info, so
```
same comment over here about the off field

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -88,11 +88,10 @@ static __always_inline bool tls_parse_field_literal(tls_dispatcher_arguments_t *
-    if (index == kIndexPath) {
-        update_path_size_telemetry(http2_tel, str_len);
-    } else {
+    if (index != kIndexPath && index != kEmptyPath) {
```
please add doc that explains this, it is not so intuitive to understand it

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -157,25 +162,25 @@ static __always_inline __u8 tls_filter_relevant_headers(tls_dispatcher_arguments
+        max_bits = is_indexed ? MAX_7_BITS : max_bits;
+        // else, if we're in a literal header - the max bits are 6.
+        max_bits = is_literal ? MAX_6_BITS : max_bits;
+        // otherwise, we're in literal header without indexing or literal header never indexed - and for both, the
```
Can you maybe share as a comment link to the RFC which explains about this use case?

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -249,12 +255,15 @@ static __always_inline void tls_process_headers(tls_dispatcher_arguments_t *info
         } else {
-            dynamic_value.string_len = current_header->new_dynamic_value_size;
-            dynamic_value.is_huffman_encoded = current_header->is_huffman_encoded;
+            // We're in new dynamic header or new dynamic header not indexed states.
```
remove the empty line in 259

hmahmood commented on pkg/network/tracer/usm_grpc_monitor_test.go:
```diff
@@ -68,15 +60,27 @@ func TestGRPCScenarios(t *testing.T) {
-		suite.Run(t, new(USMgRPCSuite))
+	ebpftest.TestBuildModes(t, []ebpftest.BuildMode{ebpftest.Prebuilt, ebpftest.RuntimeCompiled, ebpftest.CORE}, "without TLS", func(t *testing.T) {
+		grpcSuite := &USMgRPCSuite{
+			fixtures: &plainGRPCFixtures{},
```
Where is `fixtures` referenced?

guyarb replied:
it was a bad rebase

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -517,10 +533,27 @@ int uprobe__http2_tls_handle_first_frame(struct pt_regs *ctx) {
     }
     dispatcher_args_copy.off += current_frame.length;
+    // We're exceeding the packet boundaries, so we have a remainder.
+    if (dispatcher_args_copy.off > dispatcher_args_copy.len) {
```
what about dispatcher_args_copy.off = dispatcher_args_copy.len? isn't it possible to happen?

Yumasi replied:
If `dispatcher_args_copy.off == dispatcher_args_copy.len`, that means the frame perfectly fits in the buffer, in which case we don't need to store a remainder.

amitslavin replied:
You are right. thanks

amitslavin commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -517,10 +533,27 @@ int uprobe__http2_tls_handle_first_frame(struct pt_regs *ctx) {
     // frame.
     args->off = dispatcher_args_copy.off;
-
+    // Calling the next tail call.
```
not sure this is helpful

guyarb committed:
usm: tls: Skip TestLargeBodiesGRPCScenarios for tls

Currently we're still flaky around the edge cases we're testing in TestLargeBodiesGRPCScenarios.

guyarb committed:
Merge branch 'main' into guillaume.pagnoux/USMON-514-grpc-tls-v2

Yumasi committed:
Fix CR first round

Signed-off-by: Guillaume Pagnoux <guillaume.pagnoux@datadoghq.com>

guyarb committed:
Merge branch 'main' into guillaume.pagnoux/USMON-514-grpc-tls-v2

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-tls.h:
```diff
@@ -0,0 +1,459 @@
+// - HEADERS frames
+// - RST_STREAM frames
+// - DATA frames with the END_STREAM flag set
+static __always_inline void tls_find_relevant_frames(tls_dispatcher_arguments_t *info, http2_tail_call_state_t *iteration_value, http2_telemetry_t *http2_tel) {
```
that's true. We'll keep it as a separate PR as it is not mandatory for the initial version

guyarb replied:
https://datadoghq.atlassian.net/browse/USMON-684

guyarb commented on pkg/network/ebpf/c/protocols/classification/structs.h:
```diff
@@ -27,6 +27,8 @@ typedef struct {
     conn_tuple_t tup;
     __u64 tags;
     char *buffer_ptr;
+    size_t len;
```
https://datadoghq.atlassian.net/browse/USMON-685
